### PR TITLE
[Draft] Shield Funds from cached UTXOs

### PR DESCRIPTION
**Pull request by pacu**
The Idea of this PR is start to improve this PoC to the point it can be merged into the rust Crates.

Background: During the Holiday break I had the idea of putting together a Proof of Concept of how we could manage to shield funds from a t-address a light client using the data-access-api branch (now on master) and adding the needed functionality sot that:

given a set of UTXOs for a t-address obtained from lightwalletd, a light client could store them in a cache (BlockDb) and invoke a function in librustzcash that would take those funds and shield them to a z-address derived from a spending key. The user would have to use the Transparent Secret Key corresponding to those transparent outputs as proof of _spendability_

Librustzcash should be able to get all the UTXOs from the shared cache and build a shielding transaction

### CODE REVIEW

File: zcash_client_sqlite/src/chain.rs [hunk 1]
```diff
@@ -1,16 +1,80 @@
...
+    proto::compact_formats::CompactBlock,
+    address::RecipientAddress,
+    };
 
 use crate::{error::SqliteClientError, BlockDB};
 
 pub mod init;
 
+
+pub struct UnspentTransactionOutput {
```
Comment (nuttycom) [inline zcash_client_sqlite/src/chain.rs:22]: This type should probably go in zcash-client-backend; we don't actually want any functions depending on any types or calling anything in zcash_client_sqlite directly; the only things that should be used from this crate (not through the DAA) are the WalletDB and BlockDB initialization functions.

File: zcash_client_sqlite/src/chain.rs [hunk 2]
```diff
@@ -1,16 +1,80 @@
...
+    pub height: BlockHeight,
+}
+
+pub fn get_confirmed_utxos_for_address<P: consensus::Parameters>(
+    params: &P,
+    cache: &BlockDB,
+    anchor_height: BlockHeight,
+    address: &str 
+) -> Result<Vec<UnspentTransactionOutput>,SqliteClientError> {
+    let mut stmt_blocks = cache.0.prepare(
```
Comment (nuttycom) [inline zcash_client_sqlite/src/chain.rs:37]: What is inserting data into this table?

File: zcash_client_sqlite/src/chain.rs [hunk 3]
```diff
@@ -1,16 +1,80 @@
...
+    pub script: Vec<u8>,
+    pub value: Amount,
+    pub height: BlockHeight,
+}
+
+pub fn get_confirmed_utxos_for_address<P: consensus::Parameters>(
+    params: &P,
+    cache: &BlockDB,
+    anchor_height: BlockHeight,
+    address: &str
```
Comment (nuttycom) [inline zcash_client_sqlite/src/chain.rs:35]: Can this address be wrapped in a better type? You could pass the Magna Carta here. :)

File: zcash_client_sqlite/src/chain.rs [hunk 4]
```diff
@@ -1,16 +1,80 @@
...
+
+    let rows = stmt_blocks.query_map(
+        params![address, u32::from(anchor_height)],
+        |row| {
+            let addr: String = row.get(0)?;
+            let address = RecipientAddress::decode(params, &addr)
+            .   ok_or(format!(
+                "Could not interpret {} as a valid Zcash address.",
+                addr
+            )).unwrap();
```
Comment (nuttycom) [inline zcash_client_sqlite/src/chain.rs:49]: unwrap() should never be used, this error needs to be propagated out. Here's an example of something similar: https://github.com/zcash/librustzcash/blob/master/zcash_client_sqlite/src/wallet.rs#L79-L85

File: zcash_client_sqlite/src/chain.rs [hunk 5]
```diff
@@ -1,16 +1,80 @@
...
+            let script: Vec<u8> = row.get(3)?;
+            let value: i64 = row.get(4)?;
+            let height: u32 = row.get(5)?;
+
+            Ok(UnspentTransactionOutput{
+                address: address,
+                txid: txid,
+                index: index,
+                script: script,
+                value: Amount::from_i64(value).unwrap(),
```
Comment (nuttycom) [inline zcash_client_sqlite/src/chain.rs:64]: Same comment about unwrap() here, you don't want to crash the process.

File: zcash_client_sqlite/src/chain.rs [hunk 6]
```diff
@@ -1,16 +1,80 @@
...
+                height: BlockHeight::from(height)
+            })
+        },
+    )?;
+
+    let mut utxos = Vec::<UnspentTransactionOutput>::new();
+
+    for utxo in rows {
+        utxos.push(utxo.unwrap())
```
Comment (nuttycom) [inline zcash_client_sqlite/src/chain.rs:73]: Same, no unwrap, this is a place to use collect()

**Review by nuttycom — CHANGES_REQUESTED** [review]
I'm happy to pair with you to fix up any/all the issues so far! But I have a larger question, shouldn't the actual construction of the autoshielding transaction be done using the transaction builder in the zcash_client_backend trait? Right now you must be doing that outside of the rust code, right?

File: zcash_client_sqlite/src/chain.rs (hunk 2 above)
Comment (pacu) [reply zcash_client_sqlite/src/chain.rs:37]: The rust welding on mobile wallets. The idea is that libruszcash only reads from this table as it only reads the compact_blocks table. Should I add the insertion code as well to make this 'self-sufficient' ? what's the compact_blocks approach?

**Comment by nuttycom** [comment]
Subsumed by #341

**Pull request closed without merging**